### PR TITLE
Adding Circuit Breaker proposal.

### PR DESCRIPTION
**Pull request by johnament**
# Proposal for a MicroProfile specification

**! DO NOT MERGE this Pull Request BEFORE REVIEW is finished!**

**Merging this Pull Request means that the proposal is officially accepted by the MicroProfile Core Team as a MicroProfile incubation project. A public discussion and an agreement within the MicroProfile Core Team is expected before a proposal is accepted/rejected.** 

_If this PR is not a proposal for a new specification, delete this template description._

## Suggested steps to follow before this proposal is accepted ##

 - add label `proposal` to this PR on github
 - initiate a public discussion on the mP mailing list
   - Email Subject: "Review announcement: <<PROPOSAL_NAME>>"
   - Email Body: Based on [review announcement template](https://github.com/microprofile/evolution/blob/master/process.md#review-announcement) 
       - fill in `<<PROPOSAL NAME>>`, `<<REVIEW END DATE>>` (not mandatory, can be announced later, but should be at least 7 days after the announcement email is sent) 
       and `<<REVIEW MANAGER NAME>>` (the name of the sender)
       - update the link to a proposal (NNNN-proposal.md) to point to this Pull Request

After the review, the Core Team should form an agreement whether or not to accept the proposal, or whether ask its author for amendments. 

Once this Pull Reuest is merged, the proposal is considered accepted and an incubation repository should be created to facilitate collaboration.

### CODE REVIEW

**Comment by Ladicek** [comment]
How are timeouts and bulkheads expected to be solved? IMHO, circuit breakers without timeouts and bulkheads don't make a lot of sense.

**Comment by OndroMih** [comment]
I assume that with synchronous API, timeouts would be just another exception, causing a retry or a failure, would that be enough?

**Comment by OndroMih** [comment]
I would also think about asynchronous API, which is more important when calling an external service. E.g., Hystrix implements asynchronous API first ([Hystrix Flow](https://github.com/Netflix/Hystrix/wiki/How-it-Works#Flow)), and synchronous is built on top of it.

File: proposals/0002-circuit-breaker.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+## Motivation
+
+Adding circuit breakers at the platform level when dealing with microservices feels like a given, it seems like a common part of the SDK that simply should be there.  Adding it into Microprofile allows us to innovate that SDK and solicit input.
+
+## Proposed solution
+
+It would be specified that each implementation provides a CDI Interceptor for `@CheckCircuits` that allows a bean method to be demarcated as circuit breaker aware.  Failures of that method (throwing `Exception`) would cause the method to fail.  Depending on the state of the circuit, this may cause a fall back to be executed.  If the circuit is already seen as open, the method wouldn't be invoked and the fall back always invoked.
+
+The actual pattern for the circuit can be modeled one of two ways:
+- A platform inspecific definition by providing a `@Named` definition of the circuit
```
Comment (agoncal) [inline proposals/0002-circuit-breaker.md:29]: Can you explain what you mean by "platform inspecific" ?

Comment (johnament) [reply proposals/0002-circuit-breaker.md:29]: Meaning its some interface we define, not a bean class defined by a framework.  e.g. `io.microprofile.circuits.CircuitBreakerDefinition`

**Comment by johnament** [comment]
@OndrejM I'm confused... so far this doesn't expose any API.  Personally I feel like the expose of sync vs async is an implementation detail.. maybe its in scope maybe its not. It could even be inferred by the return type.  I'll play with that idea a bit more.

**Comment by OndroMih** [comment]
@johnament I agree that async might somehow be implemented by the same approach, and only the examples show a synchronous approach. I need to look at it from a bigger perspective. In current stage, we  should focus more on what are the most important aspects of circuit breakers and what we want to solve with this proposal. I'll try to provide more feedback in the mailing list, the discussion there is interesting and promising.

**Pull request closed without merging**